Frontend/full height course cards

ryan-conn opened the pull request:
## Description

Moves expand/collapse for course cards to redux, only one can be expanded at a time, and the expanded card will fill as much height as possible (with a minimum height roughly equal to what it was previously). No tests right now because I wanted some input on if we like how this works, specific behavior, etc.

## Rationale
The only thing in this branch that isn't how I imagined it is the extra height when the currently selected card is in section view and only has a couple sections. I couldn't think of any way to keep the fit-to-height behavior as well as having a minimum height without making this compromise (despite spending 4 hours in CSS hell), but I still think that in the vast majority of use cases this is a lot nicer.

## How to test
Make some cards, the expanded one will fit to your screen size (unless you're Adel and have a screen made for ants, in which case I've made it so this works basically the same way as before)

## Screenshots
I think that 7 course cards is about as high as most users will go, so here are some example sizes:

| Old appearance (always 300px) | 1080, 5 cards (417px) | 1080p, 7 cards (340px) |
|--|--|--|
| ![image](https://user-images.githubusercontent.com/28655462/95519833-8d42c700-098b-11eb-92fa-19a1d77e1638.png) | ![image](https://user-images.githubusercontent.com/28655462/95518834-861ab980-0989-11eb-8fb5-353c052de7f3.png) | ![image](https://user-images.githubusercontent.com/28655462/95518794-713e2600-0989-11eb-876b-42cf7451c5f1.png) |

| 1440p, 7 cards (697px) | 720p (314px, any more than 3 cards makes it scroll) |
|--|--|
| ![image](https://user-images.githubusercontent.com/28655462/95519611-0db4f800-098b-11eb-8c24-0172b8f01265.png) | ![image](https://user-images.githubusercontent.com/28655462/95520319-9f713500-098c-11eb-900a-69384691fa63.png) |

Finally, here's the one case I don't really like the appearance of. If you're on the section select and there are only 1 or 2 sections, there will be some extra empty space that I couldn't find any way to avoid even using refs:
![image](https://user-images.githubusercontent.com/28655462/95520472-fc6ceb00-098c-11eb-95c3-ce3087ff9873.png)
It looks normal on basic select though since I check if you're on section select with a course selected before giving a minimum height:
![image](https://user-images.githubusercontent.com/28655462/95520530-1b6b7d00-098d-11eb-8fbb-479b0f5a80c6.png)

## Related tasks

Closes #377.

ryan-conn commented:
Should be ready for review now that the edge case with low height section select has been fixed and I added tests.

firejake308 commented:
I want this merged because I am so tired of manually collapsing course cards

gannonprudhomme commented:
> I want this merged because I am so tired of manually collapsing course cards

Then review it 😉

ryan-conn commented:
Changes I just pushed make it so if you delete the expanded card, it will expand one next to it (below if possible, or above if none are below it)

gannonprudhomme commented:
Bug I ran into:
![full-course-cards-bug](https://user-images.githubusercontent.com/7295783/97513435-46c50480-195a-11eb-9f59-c0fa5e64d97c.gif)

ryan-conn commented:
> Bug I ran into:

Fixed, this was due to destructuring the existing state instead of the new card attributes when `addCourseCard` is called with an existing card.

I also noticed making this change that it will *always* expand the new card (even when loading your cards, when you don't want to expand them), so I fixed that as well

firejake308 commented on autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectColumn.tsx:
```diff
@@ -19,10 +20,20 @@ const CourseSelectColumn: React.FC = () => {
+  React.useLayoutEffect(() => {
+    if (expandedRowRef.current) {
+      const expandedRowHeight = expandedRowRef.current.children[0].clientHeight;
+      if (expandedRowHeight < 500 - 8) {
```
We should probably mention that 500 needs to be the minimum height from the `.expanded-row` rule, and that the 8 comes from the `padding-top` on `.row`

firejake308 commented on autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectColumn.css:
```diff
@@ -13,25 +14,35 @@
 
-.row:nth-last-of-type(2) {
-    padding-bottom: 0px;
+.expanded-row-overflow {
```
I would rename this to `expanded-row-small` since it's applied to small/short cards that fit within the 500px limit

firejake308 reviewed (changes requested):
Tried to break the code in browser by mindlessly adding, deleting, expanding, and collapsing different course cards, no bugs found. Love `autoFocus` btw. 2 suggestions for comments/code style and this should be good to go. We should probably have Gannon test it too, though, since he's better at teasing out bugs than I am usually.

gannonprudhomme reviewed (commented):
minor things I wrote a bit ago, will re-review later tn

ryan-conn commented:
Comments should be addressed, updated with master

gannonprudhomme commented on autoscheduler/frontend/src/redux/reducers/courseCards.ts:
```diff
@@ -43,34 +43,97 @@ const initialCourseCardArray: CourseCardArray = {
+      newState = { ...newState, [action.index]: undefined };
+      return newState;
+    }
+    case UPDATE_COURSE_CARD: {
+      // if card doesn't exist, don't update
       if (!state[action.index]) return state;
+      // if card was expanded, collapse other cards
+      if (action.courseCard.collapsed === false && state[action.index]?.collapsed !== false) {
+        return getStateAfterExpanding(state, action.index, action.courseCard);
+      }
+
+      // otherwise just update this card
       return {
         ...state,
         [action.index]: { ...state[action.index], ...action.courseCard },
         numCardsCreated: Math.max(state.numCardsCreated, action.index + 1),
       };
+    }
```
Don't think you need the braces around this case block, as you don't declare anything

gannonprudhomme reviewed (changes requested):
Few minor things, otherwise this looks good to me; also couldn't find any bug when I tried messing with it
